lib/server/roster.ts

```diff
@@ -0,0 +1,221 @@
+
+export async function getRosterArtifacts(
+  userId: string,
+): Promise<ArtifactData[]> {
+  const artifactData = (await drizzle
+    .select({
+      artifactId: rosterArtifacts.artifactId,
+      level: rosterArtifacts.level,
+    })
+    .from(rosterArtifacts)
+    .leftJoin(roster, eq(roster.id, rosterArtifacts.rosterId))
+    .where(eq(roster.userId, userId))
+    .orderBy(rosterArtifacts.artifactId)) as ArtifactData[];
+
+  return artifactData;
+}
+
+export type LevelData = {
+  levelId: number;
+  level: number;
+};
+
+export async function createOrUpdateLevel(
+  rosterId: number,
+  level: LevelData,
+): Promise<string | false> {
+  const values = {
+    rosterId: rosterId,
+    levelId: level.levelId,
+    level: level.level,
+  };
+  const createOrUpdateResponse = await drizzle
+    .insert(rosterLevels)
+    .values(values)
+    .onConflictDoUpdate({
+      target: [rosterLevels.rosterId, rosterLevels.levelId],
+      set: { level: values.level },
+    })
+    .returning({ id: rosterLevels.id });
+
+  return createOrUpdateResponse
+    ? createOrUpdateResponse[0].id.toString()
+    : false;
+}
+
+export async function createOrUpdateLevels(
+  levels: LevelData[],
+): Promise<string | false> {
+  const { userId } = auth();
+
+  let rosterId: any;
+  if (userId) {
+    rosterId = await createOrUpdateRoster(userId);
+    levels.forEach((level: any) => {
+      createOrUpdateLevel(rosterId, level);
+    });
+  } else {
+    return false;
+  }
+
+  return rosterId;
+}
+
+export async function getRosterLevels(userId: string): Promise<LevelData[]> {
+  const levelData = (await drizzle
+    .select({
+      levelId: rosterLevels.levelId,
+      level: rosterLevels.level,
+    })
+    .from(rosterLevels)
+    .leftJoin(roster, eq(roster.id, rosterLevels.rosterId))
+    .where(eq(roster.userId, userId))
+    .orderBy(rosterLevels.levelId)) as LevelData[];
+
+  return levelData;
+}
+
+export type HeroData = {
+  heroId: number;
+  ascension: AscensionLevel;
+  equipment: number;
+};
+
+export async function createOrUpdateHero(
+  rosterId: number,
+  hero: HeroData,
+): Promise<string | false> {
+  const values = {
+    rosterId: rosterId,
+    heroId: hero.heroId,
+    ascension: hero.ascension,
+    equipment: hero.equipment,
+  };
+
+  const createOrUpdateResponse = await drizzle
+    .insert(rosterHeroes)
+    .values(values)
+    .onConflictDoUpdate({
+      target: [rosterHeroes.rosterId, rosterHeroes.heroId],
+      set: { ascension: values.ascension, equipment: values.equipment },
+    })
+    .returning({ id: rosterHeroes.id });
+
+  return createOrUpdateResponse
+    ? createOrUpdateResponse[0].id.toString()
+    : false;
+}
+
+export async function createOrUpdateHeroes(
+  heroes: HeroData[],
+): Promise<string | false> {
+  const { userId } = auth();
+
+  let rosterId: any;
+  if (userId) {
+    rosterId = await createOrUpdateRoster(userId);
+    heroes.forEach((hero: any) => {
+      createOrUpdateHero(rosterId, hero);
+    });
```

Comment:
`let` and `Promise.all`

Reply:
Not sure I follow this? Are you saying rosterId shouldn't be `let` and not `any` type?

Reply:
https://github.com/KMontag42/afk-journey-team-builder/pull/48#discussion_r1744756659